### PR TITLE
update GEP5 version policy documentation

### PR DESCRIPTION
**Pull request by danielfoehrKn**
**How to categorize this PR?**

/area documentation
/kind enhancement
/priority normal

**What this PR does / why we need it**:

Add documentation for GEP 5 - Version Policy

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator

```

### CODE REVIEW

File: docs/usage/shoot_versions.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
+# Kubernetes and Machine image versions in Gardener
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md]: Can you add the document to the index (docs/README.md)?

File: docs/usage/shoot_versions.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+## Motivation
+
+On the one hand-side Gardener is responsible for managing the Kubernetes and Machine image versions of its Shoot clusters.
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md]:
```suggestion
On the one hand-side, Gardener is responsible for managing the Kubernetes and Machine image versions of its Shoot clusters.
```

File: docs/usage/shoot_versions.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+On the other hand-side, Gardener needs to be configured and updated based on the availability and support of the Kubernetes and Machine image version it provides.
+
+For instance the Kubernetes community releases **minor** versions roughly every three months and usually maintains **three minor** versions (the actual and the last two) with bug fixes and security updates.
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md]:
```suggestion
For instance, the Kubernetes community releases **minor** versions roughly every three months and usually maintains **three minor** versions (the current and the last two) with bug fixes and security updates. 
```

File: docs/usage/shoot_versions.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+## Motivation
+
+On the one hand-side Gardener is responsible for managing the Kubernetes and Machine image versions of its Shoot clusters.
+On the other hand-side, Gardener needs to be configured and updated based on the availability and support of the Kubernetes and Machine image version it provides.
+
+For instance the Kubernetes community releases **minor** versions roughly every three months and usually maintains **three minor** versions (the actual and the last two) with bug fixes and security updates. 
+Patch releases are done more frequently.
+
+Gardener needs to be configured accordingly to reflect the "logical state" of a version. 
+It should be possible to define the Kubernetes or Machine image versions that still receive bug fixes and security patches and also vice-versa to define the version that are out-of-maintenance and are potentially vulnerable.
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md]:
```suggestion
It should be possible to define the Kubernetes or Machine image versions that still receive bug fixes and security patches, and also vice-versa to define the version that are out-of-maintenance and are potentially vulnerable.
```

File: docs/usage/shoot_versions.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+Moreover, this allows Gardener to "understand" the current state of a version and act upon it (more information in the following sections). 
+
+## Overview
+
+**As a Gardener operator**:
+- I can classify a version based on it's logical state (`preview`, `supported`, `deprecated` and `expired` see [Version Classification](#version-classifications)).
+- I can define which Machine image and Kubernetes versions are eligible for the auto update of Clusters during the maintenance time.
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md]:
```suggestion
- I can define which Machine image and Kubernetes versions are eligible for the auto update of clusters during the maintenance time.
```

File: docs/usage/shoot_versions.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+- I can disallow the creation of clusters having a certain version (think of severe security issues).
+
+**As an end-user/Shoot owner of Gardener**:
+- I can get information which Kubernetes and Machine image versions are supported, in preview, deprecated or expired.
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md]:
```suggestion
- I can get information about which Kubernetes and Machine image versions exist and their classification. 
```

File: docs/usage/shoot_versions.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+- I can disallow the creation of clusters having a certain version (think of severe security issues).
+
+**As an end-user/Shoot owner of Gardener**:
+- I can get information which Kubernetes and Machine image versions are supported, in preview, deprecated or expired. 
+- I can determine the time when my Shoot clusters Machine image and Kubernetes version will be forcefully updated to the next patch or minor version (in case the cluster is running a deprecated version with an expiration date).
+- I can get this information via API from the CloudProfile.
+
+## Version Classifications
+
+In their lifetime, versions go through 4 distinct "logical states".
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md]: Can you explain that this is purely the decision of a human operator and could also be done entirely differently, i.e., that what you explain here is just "a best/recommended practice"?

File: docs/usage/shoot_versions.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+Versions usually start with the classification `preview`, then are promoted to `supported`, eventually `deprecated` and finally `expired`. 
+This classification serves as a "point-of-reference" for end-users and also has implications during shoot creation and the maintenance time.
+
+This information is programmatically available in the `CloudProfiles` of the Garden cluster.
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md:40]: Sometimes you escape things like `CloudProfile`, sometimes you don't.

File: docs/usage/shoot_versions.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+## Version Classifications
+
+In their lifetime, versions go through 4 distinct "logical states".
+Versions usually start with the classification `preview`, then are promoted to `supported`, eventually `deprecated` and finally `expired`. 
+This classification serves as a "point-of-reference" for end-users and also has implications during shoot creation and the maintenance time.
+
+This information is programmatically available in the `CloudProfiles` of the Garden cluster. 
+
+- **preview:** A `preview` version is a new version that has not yet undergone thorough testing, possibly a new release and needs time to be validated. There is a probability of unresolved issues and is therefore not yet recommended for production usage.
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md]:
```suggestion
- **preview:** A `preview` version is a new version that has not yet undergone thorough testing, possibly a new release, and needs time to be validated. Due to its short early age, there is a higher probability of undiscovered issues and is therefore not yet recommended for production usage.
```

File: docs/usage/shoot_versions.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+- **supported:** A `supported` version is the recommended version for new and existing Shoot clusters. New Shoot clusters should use and existing clusters should update to this version.
+Typically for Kubernetes versions, the latest Kubernetes patch versions of the actual (if not still in `preview`) and the last 3 minor Kubernetes versions are maintained by the community. An operator could define these versions as being `supported` (e.g. v1.16.1, v1.15.4, v1.14.9 and v1.13.12).
+
+- **deprecated:** A `deprecated` version is a version that approaches the end of its lifecycle and can contain issues which are probably resolved in a supported version. 
+New Shoots should not use this version any more. 
+Existing Shoots will be updated to a newer version if `auto-update` is enabled (`.spec.maintenance.autoUpdate.kubernetesVersion` for Kubernetes version `auto-update` or `.spec.maintenance.autoUpdate.machineImageVersion` for machine machine image version `auto-update`).
+Using automatic upgrades however does not guarantee that a Shoot runs a non-deprecated version, as the latest version (overall or of the minor version) can be deprecated as well.
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md]:
```suggestion
Using automatic upgrades, however, does not guarantee that a Shoot runs a non-deprecated version as the latest version (overall or of the minor version) can be deprecated as well.
```

File: docs/usage/shoot_versions.md [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+After the operator gained sufficient trust, the version can be manually promoted to `supported`.  
+
+- **supported:** A `supported` version is the recommended version for new and existing Shoot clusters. New Shoot clusters should use and existing clusters should update to this version.
+Typically for Kubernetes versions, the latest Kubernetes patch versions of the actual (if not still in `preview`) and the last 3 minor Kubernetes versions are maintained by the community. An operator could define these versions as being `supported` (e.g. v1.16.1, v1.15.4, v1.14.9 and v1.13.12).
+
+- **deprecated:** A `deprecated` version is a version that approaches the end of its lifecycle and can contain issues which are probably resolved in a supported version. 
+New Shoots should not use this version any more. 
+Existing Shoots will be updated to a newer version if `auto-update` is enabled (`.spec.maintenance.autoUpdate.kubernetesVersion` for Kubernetes version `auto-update` or `.spec.maintenance.autoUpdate.machineImageVersion` for machine machine image version `auto-update`).
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md]:
```suggestion
Existing Shoots will be updated to a newer version if `auto-update` is enabled (`.spec.maintenance.autoUpdate.kubernetesVersion` for Kubernetes version `auto-update`, or `.spec.maintenance.autoUpdate.machineImageVersion` for machine machine image version `auto-update`).
```

File: docs/usage/shoot_versions.md [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+For **Kubernetes versions**, the forceful update picks the latest `non-preview` patch version of the current minor version. 
+
+ If the cluster is already on the latest patch version and the latest patch version is also expired, 
+ it will continue with the latest patch version of the **next consecutive minor Kubernetes version**, so **it will result in an 
+ update of a minor Kubernetes version!** 
+ 
+ If that's expired as well, the update process repeats until a non-expired Kubernetes version is reached.
+  
+ **Depending on the circumstances described above, it can happen that the cluster receives multiple consecutive minor Kubernetes version updates!**
+ As a Kubernetes version update cannot skip a minor version, the CloudProfile has to be maintained properly.
```
Comment (rfranzke) [inline docs/usage/shoot_versions.md]: What does "maintain properly" exactly mean? Also, can you make more clear that such minor version upgrades don't happen altogether (i.e., we don't go from 1.10 -> 1.11 -> 1.12 in one maintenance cycle, but instead do it in multiple cycles)?

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: docs/usage/shoot_versions.md [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+- **deprecated:** A `deprecated` version is a version that approaches the end of its lifecycle and can contain issues which are probably resolved in a supported version. 
+New Shoots should not use this version any more. 
+Existing Shoots will be updated to a newer version if `auto-update` is enabled (`.spec.maintenance.autoUpdate.kubernetesVersion` for Kubernetes version `auto-update` or `.spec.maintenance.autoUpdate.machineImageVersion` for machine machine image version `auto-update`).
+Using automatic upgrades however does not guarantee that a Shoot runs a non-deprecated version, as the latest version (overall or of the minor version) can be deprecated as well.
+Deprecated versions should have an expiration date set for eventual expiration. 
+
+- **expired:** An `expired` versions has an expiration date in the past. 
+ New clusters with that version cannot be created and existing clusters are forcefully migrated to a higher version during the maintenance time.
+ 
+Below is an example how the relevant section of the CloudProfile might look like:
```
Comment (vpnachev) [inline docs/usage/shoot_versions.md]: Please, add a reference date with which the `expirationDate` can be compared.

File: docs/usage/shoot_maintenance.md [hunk 14]
```diff
@@ -40,8 +40,19 @@ spec:
       machineImageVersion: true
 ```
 
-Please refer to [this document](../proposals/05-versioning-policy.md) that describes the exact versioning policy.
+During the daily maintenance, the Gardener Controller Manager updates the Shoot's Kubernetes and machine image version based on the following criteria:
+ - there is a higher version available and the Shoot opted-in for automatic version updates
+ - the currently used version is `expired`
```
Comment (vpnachev) [inline docs/usage/shoot_maintenance.md:45]: Could you make it more clear that only one of the criteria is enough to trigger version update?

File: docs/usage/shoot_versions.md [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+## Motivation
+
+On the one hand-side Gardener is responsible for managing the Kubernetes and Machine image versions of its Shoot clusters.
+On the other hand-side, Gardener needs to be configured and updated based on the availability and support of the Kubernetes and Machine image version it provides.
+
+For instance the Kubernetes community releases **minor** versions roughly every three months and usually maintains **three minor** versions (the actual and the last two) with bug fixes and security updates. 
+Patch releases are done more frequently.
+
+Gardener needs to be configured accordingly to reflect the "logical state" of a version.
```
Comment (vpnachev) [inline docs/usage/shoot_versions.md]: Could you point which is the default class for a version if omitted?

**Review by vpnachev — COMMENTED** [review]
Overall looks great.

I have just few minor nits

File: docs/usage/shoot_versions.md (hunk 15 above)
Comment (danielfoehrKn) [reply docs/usage/shoot_versions.md]: I added a paragraph below  in ## Version Classifications

File: docs/usage/shoot_versions.md (hunk 13 above)
Comment (danielfoehrKn) [reply docs/usage/shoot_versions.md]: I added a reference to the Golang time package when introducing the expiration date.
The package contains extensive documentation.

File: docs/usage/shoot_versions.md [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+it will continue with the latest patch version of the **next consecutive minor Kubernetes version**, so **it will result in an 
+update of a minor Kubernetes version!** 
+ 
+Please note, that multiple consecutive minor version upgrades are possible. 
+This can occur if the Shoot is updated to a version that in turn is also `expired`. 
+In this case, the version is again upgraded in the **next** maintenance time.
+  
+**Depending on the circumstances described above, it can happen that the cluster receives multiple consecutive minor Kubernetes version updates!**
+
+Kubernetes "minor version jumps" are not allowed - meaning to skip the update to the consecutive minor version but directly updating to any version after that.
```
Comment (plkokanov) [inline docs/usage/shoot_versions.md]:
```suggestion
Kubernetes "minor version jumps" are not allowed - meaning to skip the update to the consecutive minor version and directly update to any version after that.
```

**Review by plkokanov — COMMENTED** [review]
Looks really good!
Just one very minor suggestion as it sounds a bit better to me.

**Review by vpnachev — DISMISSED** [review]
/lgtm

/needs second-opinion

**Comment by ghost** [comment]
@vpnachev Label "need-second-opinion" does not exist for this repository.

**Comment by ghost** [comment]
@vpnachev Label "needs-second-opinion" does not exist for this repository.

File: docs/usage/shoot_maintenance.md [hunk 17]
```diff
@@ -30,7 +30,7 @@ You can change it later, of course.
 ## Automatic Version Updates
 
 The `.spec.maintenance.autoUpdate` field in the shoot specification allows you to control how/whether automatic updates of Kubernetes patch and machine image versions are performed.
-During the maintenance, Gardener checks if there are newer versions available, and if you opt-in, then your shoot will be updated accordingly.
+Machine image version are updated per worker pool.
```
Comment (rfranzke) [inline docs/usage/shoot_maintenance.md]:
```suggestion
Machine image versions are updated per worker pool.
```

**Review by rfranzke — DISMISSED** [review]
/lgtm

File: docs/usage/shoot_maintenance.md (hunk 17 above)
Comment (vpnachev) [reply docs/usage/shoot_maintenance.md]: @danielfoehrKn could you apply this proposal?
The rest is looking good to me.

Comment (danielfoehrKn) [reply docs/usage/shoot_maintenance.md]: done

**Review by vpnachev — APPROVED** [review]

**Pull request merged**